### ISSUE TITLE
Implement the Dataframe Type

### ISSUE THREAD

**Issue by iamrecursion** [enso-org/enso#567]
### Summary
For a language intended for data cleaning, processing, analysis and manipulation, no collection type is more key to its success than the dataframe. A dataframe is a tabular representation of data, consisting of (named) columns of a given type, that can be filtered, manipulated, processed, and calculated upon. 

This is, clearly, a crucial addition to Enso, and has to be done properly. 

### Value
Every language that is used for data processing today has some concept of a dataframe, making it crucial that Enso provide an implementation. However, the Enso implementation has the potential to be significantly more performant than many existing implementations (e.g. Python's Pandas), providing a major selling point for working with huge amounts of data.

### Specification

- Create a design for implementing the new version of the `Dataframe` type.
- This design must be congruent with the future plans for the type-checker and structural typing in order to minimise breaking changes in future.
- The design must account for retaining and improving much of the functionality that already exists in Luna dataframes. 
- The backing structure for the new Enso `Dataframe` type should be based on binding Apache Arrow natively in Java for performance reasons. This will allow us to rapidly prototype and implement the collection type, as well as far more easily provide performant operations on it than if we had used C++.
- Once types come along, dataframes will be parametrised over a list of types `Dataframe [T1, T2, T3, ...]`, so the design in the dynamic case needs to be congruent with this.
- Once the design is complete, it should be rapidly prototyped into a usable state.
- The accompanying library functions should implement the minimal amount of functionality. 

### Acceptance Criteria & Test Cases

- The type has been comprehensively designed.
- The implementation is congruent with this design.
- The implementation is rigorously tested and benchmarked, both at the Enso level and the backing language (Java) level.

**Comment by iamrecursion** [comment]
No longer relevant.
